File: manpower/server/core/admin_api.py
from django.contrib.auth import get_user_model
from rest_framework import parsers, status, viewsets
from rest_framework.decorators import api_view, parser_classes, permission_classes
from rest_framework.exceptions import PermissionDenied
from rest_framework.permissions import AllowAny, IsAdminUser, IsAuthenticated
from rest_framework.response import Response
from rest_framework_simplejwt.serializers import TokenObtainPairSerializer
from rest_framework_simplejwt.views import TokenObtainPairView

from .models import (
    AboutAccordionItem,
    CMSPage,
    CareerOpening,
    Certificate,
    Client,
    ContactInquiry,
    ContactNumber,
    ContentBlock,
    Country,
    DemandList,
    FAQ,
    FooterLink,
    GalleryAlbum,
    GalleryImage,
    HeroSection,
    HomeStatistic,
    Job,
    MediaAsset,
    Membership,
    MottoStep,
    NavigationLink,
    NewsArticle,
    OfficeLocation,
    PageMeta,
    PartnershipInquiry,
    QuickLink,
    RecruitmentStep,
    RegistrationSubmission,
    Sector,
    SiteConfiguration,
    SocialLink,
    Testimonial,
    WhyChooseUsItem,
)
from .serializers import (
    AboutAccordionItemSerializer,
    CMSPageSerializer,
    CareerOpeningSerializer,
    CertificateSerializer,
    ClientSerializer,
    ContactInquirySerializer,
    ContactNumberSerializer,
    ContentBlockSerializer,
    CountrySerializer,
    DemandListSerializer,
    FAQSerializer,
    FooterLinkSerializer,
    GalleryAlbumSerializer,
    GalleryImageSerializer,
    HeroSectionSerializer,
    HomeStatisticSerializer,
    JobSerializer,
    MediaAssetSerializer,
    MembershipSerializer,
    MottoStepSerializer,
    NavigationLinkSerializer,
    NewsArticleSerializer,
    OfficeLocationSerializer,
    PageMetaSerializer,
    PartnershipInquirySerializer,
    QuickLinkSerializer,
    RecruitmentStepSerializer,
    RegistrationSubmissionSerializer,
    SectorSerializer,
    SiteConfigurationSerializer,
    SocialLinkSerializer,
    TestimonialSerializer,
    UserAdminSerializer,
    UserSerializer,
    WhyChooseUsItemSerializer,
)

User = get_user_model()


class StaffTokenObtainPairSerializer(TokenObtainPairSerializer):
    def validate(self, attrs):
        data = super().validate(attrs)
        if not self.user.is_staff:
            raise PermissionDenied(
                "Staff access required. This account cannot use Super Admin."
            )
        data["user"] = UserSerializer(self.user).data
        return data


class StaffTokenObtainPairView(TokenObtainPairView):
    serializer_class = StaffTokenObtainPairSerializer
    permission_classes = [AllowAny]


@api_view(["GET"])
@permission_classes([IsAuthenticated, IsAdminUser])
def me(request):
    return Response(UserSerializer(request.user).data)


@api_view(["GET"])
@permission_classes([IsAuthenticated, IsAdminUser])
def dashboard(request):
    return Response(
        {
            "jobs_total": Job.objects.count(),
            "jobs_active": Job.objects.filter(is_active=True).count(),
            "news_total": NewsArticle.objects.count(),
            "news_published": NewsArticle.objects.filter(is_published=True).count(),
            "sectors_total": Sector.objects.count(),
            "demands_total": DemandList.objects.filter(is_active=True).count(),
            "gallery_total": GalleryImage.objects.count(),
            "pages_total": CMSPage.objects.count(),
            "inquiries_open": (
                ContactInquiry.objects.filter(handled=False).count()
                + PartnershipInquiry.objects.filter(handled=False).count()
                + RegistrationSubmission.objects.filter(handled=False).count()
            ),
            "contact_open": ContactInquiry.objects.filter(handled=False).count(),
            "partnership_open": PartnershipInquiry.objects.filter(handled=False).count(),
            "registration_open": RegistrationSubmission.objects.filter(handled=False).count(),
            "users_total": User.objects.count(),
        }
    )


class AdminModelViewSet(viewsets.ModelViewSet):
    permission_classes = [IsAuthenticated, IsAdminUser]

    def get_serializer_context(self):
        ctx = super().get_serializer_context()
        ctx["request"] = self.request
        return ctx


class MultiPartMixin:
    parser_classes = [parsers.MultiPartParser, parsers.FormParser, parsers.JSONParser]


class JobViewSet(AdminModelViewSet):
    queryset = Job.objects.select_related("sector", "country").all()
    serializer_class = JobSerializer


class NewsArticleViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = NewsArticle.objects.all()
    serializer_class = NewsArticleSerializer


class SectorViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = Sector.objects.all()
    serializer_class = SectorSerializer


class CountryViewSet(AdminModelViewSet):
    queryset = Country.objects.all()
    serializer_class = CountrySerializer


class HeroSectionViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = HeroSection.objects.all()
    serializer_class = HeroSectionSerializer


class HomeStatisticViewSet(AdminModelViewSet):
    queryset = HomeStatistic.objects.all()
    serializer_class = HomeStatisticSerializer


class MottoStepViewSet(AdminModelViewSet):
    queryset = MottoStep.objects.all()
    serializer_class = MottoStepSerializer


class TestimonialViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = Testimonial.objects.all()
    serializer_class = TestimonialSerializer


class MembershipViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = Membership.objects.all()
    serializer_class = MembershipSerializer


class ClientViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = Client.objects.all()
    serializer_class = ClientSerializer


class WhyChooseUsItemViewSet(AdminModelViewSet):
    queryset = WhyChooseUsItem.objects.all()
    serializer_class = WhyChooseUsItemSerializer


class AboutAccordionItemViewSet(AdminModelViewSet):
    queryset = AboutAccordionItem.objects.all()
    serializer_class = AboutAccordionItemSerializer


class RecruitmentStepViewSet(AdminModelViewSet):
    queryset = RecruitmentStep.objects.all()
    serializer_class = RecruitmentStepSerializer


class CertificateViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = Certificate.objects.all()
    serializer_class = CertificateSerializer


class NavigationLinkViewSet(AdminModelViewSet):
    queryset = NavigationLink.objects.all()
    serializer_class = NavigationLinkSerializer


class FooterLinkViewSet(AdminModelViewSet):
    queryset = FooterLink.objects.all()
    serializer_class = FooterLinkSerializer


class QuickLinkViewSet(AdminModelViewSet):
    queryset = QuickLink.objects.all()
    serializer_class = QuickLinkSerializer


class FAQViewSet(AdminModelViewSet):
    queryset = FAQ.objects.all()
    serializer_class = FAQSerializer


class SocialLinkViewSet(AdminModelViewSet):
    queryset = SocialLink.objects.all()
    serializer_class = SocialLinkSerializer


class ContactNumberViewSet(AdminModelViewSet):
    queryset = ContactNumber.objects.select_related("site").all()
    serializer_class = ContactNumberSerializer


class OfficeLocationViewSet(AdminModelViewSet):
    queryset = OfficeLocation.objects.select_related("site").all()
    serializer_class = OfficeLocationSerializer


class DemandListViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = DemandList.objects.select_related("sector", "country").all()
    serializer_class = DemandListSerializer


class CareerOpeningViewSet(AdminModelViewSet):
    queryset = CareerOpening.objects.all()
    serializer_class = CareerOpeningSerializer


class GalleryAlbumViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = GalleryAlbum.objects.prefetch_related("images").all()
    serializer_class = GalleryAlbumSerializer


class GalleryImageViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = GalleryImage.objects.select_related("album").all()
    serializer_class = GalleryImageSerializer


class PageMetaViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = PageMeta.objects.all()
    serializer_class = PageMetaSerializer


class CMSPageViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = CMSPage.objects.all()
    serializer_class = CMSPageSerializer


class MediaAssetViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = MediaAsset.objects.all()
    serializer_class = MediaAssetSerializer


class ContentBlockViewSet(MultiPartMixin, AdminModelViewSet):
    queryset = ContentBlock.objects.all()
    serializer_class = ContentBlockSerializer

    def get_queryset(self):
        qs = super().get_queryset()
        section = self.request.query_params.get("section") or self.request.query_params.get(
            "page_key"
        )
        if section:
            qs = qs.filter(page=section)
        return qs


class UserViewSet(AdminModelViewSet):
    queryset = User.objects.all().order_by("-date_joined")
    serializer_class = UserAdminSerializer

    def get_queryset(self):
        qs = super().get_queryset()
        if not self.request.user.is_superuser:
            return qs.filter(id=self.request.user.id)
        return qs

    def perform_create(self, serializer):
        if not self.request.user.is_superuser:
            raise PermissionDenied("Only superusers can create staff accounts.")
        # CMS login requires is_staff; default new accounts to staff when omitted.
        is_staff = serializer.validated_data.get("is_staff", True)
        serializer.save(is_staff=is_staff)

    def perform_destroy(self, instance):
        if instance == self.request.user:
            raise PermissionDenied("You cannot delete your own account.")
        if instance.is_superuser and not self.request.user.is_superuser:
            raise PermissionDenied("Only superusers can delete superusers.")
        instance.delete()


@api_view(["GET", "PUT", "PATCH"])
@parser_classes([parsers.MultiPartParser, parsers.FormParser, parsers.JSONParser])
@permission_classes([IsAuthenticated, IsAdminUser])
def site_settings_admin(request):
    site = SiteConfiguration.objects.order_by("-updated_at").first()
    empty = {
        "id": None,
        "company_name": "",
        "short_name": "",
        "address": "",
        "primary_email": "",
        "po_box": "",
        "notice_text": "",
        "tagline": "",
        "map_embed_url": "",
        "business_hours": "",
        "cv_download_url": "",
        "license_number": "",
        "meta_title": "",
        "meta_description": "",
        "meta_keywords": "",
        "partner_cta_heading": "Partner With Us",
        "partner_cta_body": "",
        "partner_cta_button_label": "Contact Us",
        "partner_cta_button_path": "/contact",
        "hiring_enabled": True,
        "hiring_heading": "WE ARE HIRING!",
        "hiring_subheading": "Open Positions",
        "hiring_button_label": "APPLY NOW",
        "hiring_button_path": "/online-registration",
        "ethic_heading": "Ethical Recruitment",
        "ethic_eyebrow": "What we do",
        "ethic_body": "",
        "ethic_button_label": "Learn More",
        "ethic_button_path": "/ethical-recruitment",
        "motto_heading": "Our Motto",
        "motto_intro": "",
        "expertise_heading": "Our Expertise",
        "expertise_intro": "",
        "expertise_button_label": "Learn More",
        "expertise_button_path": "/services",
        "testimonials_heading": "Testimonial",
        "membership_heading": "Membership",
        "clients_heading": "PROUD TO WORK WITH",
        "partnership_form_heading": "Partner With Us",
        "partnership_form_success": "Thank you — we received your message and will reply shortly.",
        "partnership_form_submit_label": "Submit",
        "partnership_form_sending_label": "Sending…",
        "partnership_form_name_label": "Full name",
        "partnership_form_phone_label": "Phone",
        "partnership_form_email_label": "Email",
        "partnership_form_message_label": "Message",
        "logo_url": None,
        "favicon_url": None,
        "og_image_url": None,
        "stats_background_url": None,
        "testimonials_background_url": None,
        "contact_numbers": [],
        "offices": [],
        "updated_at": None,
    }

    if site is None:
        if request.method == "GET":
            return Response(empty)
        serializer = SiteConfigurationSerializer(data=request.data, context={"request": request})
        serializer.is_valid(raise_exception=True)
        site = serializer.save()
        return Response(
            SiteConfigurationSerializer(site, context={"request": request}).data,
            status=status.HTTP_201_CREATED,
        )

    if request.method == "GET":
        return Response(SiteConfigurationSerializer(site, context={"request": request}).data)

    # Support multipart for logo/favicon/background uploads; empty string clears the file.
    image_fields = {
        "logo",
        "favicon",
        "og_image",
        "stats_background",
        "testimonials_background",
    }
    raw = request.data
    data = {}
    for key in raw:
        value = raw.get(key)
        if key in image_fields and value in ("", None):
            data[key] = None
        elif key == "hiring_enabled" and isinstance(value, str):
            data[key] = value.strip().lower() in ("1", "true", "yes", "on")
        else:
            data[key] = value

    serializer = SiteConfigurationSerializer(
        site, data=data, partial=request.method == "PATCH", context={"request": request}
    )
    serializer.is_valid(raise_exception=True)
    serializer.save()
    return Response(serializer.data)


class ContactInquiryViewSet(AdminModelViewSet):
    queryset = ContactInquiry.objects.all()
    serializer_class = ContactInquirySerializer
    http_method_names = ["get", "patch", "head", "options"]


class PartnershipInquiryViewSet(AdminModelViewSet):
    queryset = PartnershipInquiry.objects.all()
    serializer_class = PartnershipInquirySerializer
    http_method_names = ["get", "patch", "head", "options"]


class RegistrationSubmissionViewSet(AdminModelViewSet):
    queryset = RegistrationSubmission.objects.all()
    serializer_class = RegistrationSubmissionSerializer
    http_method_names = ["get", "patch", "head", "options"]

File: manpower/web/src/lib/admin-api.ts
import { api } from "./api";
import type { AdminUser } from "./auth";
import { clearTokens, setTokens } from "./auth";

export type LoginResponse = {
  access: string;
  refresh: string;
  user: AdminUser;
};

export async function login(username: string, password: string): Promise<LoginResponse> {
  const data = await api<LoginResponse>("/auth/token/", {
    method: "POST",
    body: { username, password },
    auth: false,
  });
  setTokens(data.access, data.refresh);
  return data;
}

export async function fetchMe(): Promise<AdminUser> {
  return api<AdminUser>("/auth/me/");
}

export function logout() {
  clearTokens();
}

export type DashboardStats = {
  jobs_total: number;
  jobs_active: number;
  news_total: number;
  news_published: number;
  sectors_total: number;
  demands_total: number;
  gallery_total: number;
  pages_total: number;
  inquiries_open: number;
  contact_open: number;
  partnership_open: number;
  registration_open: number;
  users_total: number;
};

export type Job = {
  id: number;
  title: string;
  company: string;
  sector: number;
  sector_name: string;
  country: number;
  country_name: string;
  vacancies: number;
  salary: string;
  description: string;
  requirements: string;
  deadline: string | null;
  is_featured: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type NewsArticle = {
  id: number;
  title: string;
  slug: string;
  author: string;
  excerpt: string;
  content: string;
  is_published: boolean;
  published_at: string | null;
  meta_title: string;
  meta_description: string;
  created_at: string;
  updated_at: string;
};

export type Sector = {
  id: number;
  name: string;
  slug: string;
  description: string;
  icon?: string;
  image_url?: string | null;
  is_featured: boolean;
  is_active: boolean;
  order: number;
};

export type Country = {
  id: number;
  name: string;
  flag_emoji: string;
  is_active: boolean;
  order: number;
};

export type SiteSettings = {
  id: number | null;
  company_name: string;
  short_name: string;
  address: string;
  primary_email: string;
  po_box: string;
  notice_text: string;
  tagline: string;
  map_embed_url: string;
  business_hours: string;
  cv_download_url: string;
  license_number: string;
  meta_title: string;
  meta_description: string;
  meta_keywords: string;
  partner_cta_heading: string;
  partner_cta_body: string;
  partner_cta_button_label: string;
  partner_cta_button_path: string;
  hiring_enabled: boolean;
  hiring_heading: string;
  hiring_subheading: string;
  hiring_button_label: string;
  hiring_button_path: string;
  ethic_heading: string;
  ethic_eyebrow: string;
  ethic_body: string;
  ethic_button_label: string;
  ethic_button_path: string;
  motto_heading?: string;
  motto_intro?: string;
  expertise_heading?: string;
  expertise_intro?: string;
  expertise_button_label?: string;
  expertise_button_path?: string;
  testimonials_heading?: string;
  membership_heading?: string;
  clients_heading?: string;
  partnership_form_heading?: string;
  partnership_form_success?: string;
  partnership_form_submit_label?: string;
  partnership_form_sending_label?: string;
  partnership_form_name_label?: string;
  partnership_form_phone_label?: string;
  partnership_form_email_label?: string;
  partnership_form_message_label?: string;
  logo_url?: string | null;
  favicon_url?: string | null;
  og_image_url?: string | null;
  stats_background_url?: string | null;
  testimonials_background_url?: string | null;
  updated_at: string | null;
};

export type ContactInquiry = {
  id: number;
  full_name: string;
  email: string;
  subject: string;
  message: string;
  phone: string;
  handled: boolean;
  created_at: string;
};

export type PartnershipInquiry = {
  id: number;
  full_name: string;
  email: string;
  phone: string;
  message: string;
  handled: boolean;
  created_at: string;
};

export type RegistrationSubmission = {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  contact_number: string;
  permanent_address: string;
  temporary_address: string;
  position: string;
  preferred_country: string;
  message: string;
  handled: boolean;
  created_at: string;
};

export type StaffUser = {
  id: number;
  username: string;
  email: string;
  first_name: string;
  last_name: string;
  is_staff: boolean;
  is_superuser: boolean;
  is_active: boolean;
  date_joined?: string;
  password?: string;
};

type CrudResource<T> = {
  list: (params?: Record<string, string>) => Promise<T[]>;
  get: (id: number) => Promise<T>;
  create: (body: Partial<T> | FormData) => Promise<T>;
  update: (id: number, body: Partial<T> | FormData) => Promise<T>;
  remove: (id: number) => Promise<void>;
};

function crud<T>(base: string): CrudResource<T> {
  return {
    list: (params) => {
      const q = params ? new URLSearchParams(params).toString() : "";
      return api<T[]>(`${base}/${q ? `?${q}` : ""}`);
    },
    get: (id) => api<T>(`${base}/${id}/`),
    create: (body) => api<T>(`${base}/`, { method: "POST", body }),
    update: (id, body) => api<T>(`${base}/${id}/`, { method: "PATCH", body }),
    remove: (id) => api<void>(`${base}/${id}/`, { method: "DELETE" }),
  };
}

export type OrderedItem = {
  id: number;
  order: number;
  is_active: boolean;
  [key: string]: unknown;
};

export const adminApi = {
  dashboard: () => api<DashboardStats>("/admin/dashboard/"),
  jobs: crud<Job>("/admin/jobs"),
  news: crud<NewsArticle>("/admin/news"),
  sectors: crud<Sector>("/admin/sectors"),
  countries: crud<Country>("/admin/countries"),
  hero: crud<OrderedItem>("/admin/hero"),
  stats: crud<OrderedItem>("/admin/stats"),
  motto: crud<OrderedItem>("/admin/motto"),
  testimonials: crud<OrderedItem>("/admin/testimonials"),
  memberships: crud<OrderedItem>("/admin/memberships"),
  clients: crud<OrderedItem>("/admin/clients"),
  whyChooseUs: crud<OrderedItem>("/admin/why-choose-us"),
  aboutAccordion: crud<OrderedItem>("/admin/about-accordion"),
  recruitmentSteps: crud<OrderedItem>("/admin/recruitment-steps"),
  certificates: crud<OrderedItem>("/admin/certificates"),
  navigation: crud<OrderedItem>("/admin/navigation"),
  footerLinks: crud<OrderedItem>("/admin/footer-links"),
  quickLinks: crud<OrderedItem>("/admin/quick-links"),
  faqs: crud<OrderedItem>("/admin/faqs"),
  socialLinks: crud<OrderedItem>("/admin/social-links"),
  contactNumbers: crud<OrderedItem>("/admin/contact-numbers"),
  offices: crud<OrderedItem>("/admin/offices"),
  demands: crud<OrderedItem>("/admin/demands"),
  careers: crud<OrderedItem>("/admin/careers"),
  galleryAlbums: crud<OrderedItem>("/admin/gallery/albums"),
  galleryImages: crud<OrderedItem>("/admin/gallery/images"),
  pageSeo: crud<OrderedItem>("/admin/page-seo"),
  pages: crud<OrderedItem>("/admin/pages"),
  contentBlocks: crud<OrderedItem>("/admin/content-blocks"),
  media: crud<OrderedItem>("/admin/media"),
  users: crud<StaffUser>("/admin/users"),
  siteSettings: {
    get: () => api<SiteSettings>("/admin/site-settings/"),
    update: (body: Partial<SiteSettings> | FormData, create = false) =>
      api<SiteSettings>("/admin/site-settings/", {
        method: create ? "PUT" : "PATCH",
        body,
      }),
  },
  inquiries: {
    contact: {
      list: () => api<ContactInquiry[]>("/admin/inquiries/contact/"),
      patch: (id: number, body: { handled: boolean }) =>
        api<ContactInquiry>(`/admin/inquiries/contact/${id}/`, { method: "PATCH", body }),
    },
    partnership: {
      list: () => api<PartnershipInquiry[]>("/admin/inquiries/partnership/"),
      patch: (id: number, body: { handled: boolean }) =>
        api<PartnershipInquiry>(`/admin/inquiries/partnership/${id}/`, {
          method: "PATCH",
          body,
        }),
    },
    registration: {
      list: () => api<RegistrationSubmission[]>("/admin/inquiries/registration/"),
      patch: (id: number, body: { handled: boolean }) =>
        api<RegistrationSubmission>(`/admin/inquiries/registration/${id}/`, {
          method: "PATCH",
          body,
        }),
    },
  },
};
